protocol/contracts/beanstalk/sun/SeasonFacet/Receiving.sol

```diff
@@ -0,0 +1,119 @@
+ * @notice Holds the functions responsible for receiving Bean shipments after mints. These
+ *  functions must live inside of the Beanstalk Diamond. If new receiving components are needed,
+ *  this contract and its containing Facet will need to be updated.
+ * @dev An alternative design could remove the need for the generalized receive() entry function
+ *  and instead require the shipping route to define the selector of its own corresponding receive
+ *  function. However, both designs will require a Facet cut if a new receive function is needed,
+ *  so this design was chosen for additional clarity.
+ */
+contract Receiving is ReentrancyGuard {
+    using SafeCast for uint256;
+
+    /**
+     * @notice General entry point to receive Beans at a given component of the system.
+     * @param recipient The Beanstalk component that will receive the Beans.
+     * @param amount The amount of Beans to receive.
+     * @param data Additional data to pass to the receiving function.
+     */
+    function receiveShipment(
+        Storage.ShipmentRecipient recipient,
+        uint256 amount,
+        bytes memory data
+    ) internal {
+        if (recipient == Storage.ShipmentRecipient.Silo) {
+            siloReceive(amount, data);
+        } else if (recipient == Storage.ShipmentRecipient.Field) {
+            fieldReceive(amount, data);
+        } else if (recipient == Storage.ShipmentRecipient.Barn) {
+            barnReceive(amount, data);
+        }
+    }
+
+    /**
+     * @notice Receive Beans at the Silo, distributing Stalk & Earned Beans.
+     * @dev Data param not used.
+     * @param amount Amount of Beans to receive.
+     */
+    function siloReceive(uint256 amount, bytes memory) private {
+        // `s.earnedBeans` is an accounting mechanism that tracks the total number
+        // of Earned Beans that are claimable by Stalkholders. When claimed via `plant()`,
+        // it is decremented. See {Silo.sol:_plant} for more details.
+        s.earnedBeans = s.earnedBeans + amount.toUint128();
+
+        // Mint Stalk (as Earned Stalk).
+        // Stalk is created here because only Beans that are allocated to the Silo receive Stalk.
+        s.silo.stalk = s.silo.stalk + (amount * C.STALK_PER_BEAN);
+
+        // SafeCast unnecessary here because of prior safe cast.
+        s.siloBalances[C.BEAN].deposited = s.siloBalances[C.BEAN].deposited + uint128(amount);
+        s.siloBalances[C.BEAN].depositedBdv = s.siloBalances[C.BEAN].depositedBdv + uint128(amount);
+    }
+
+    /**
+     * @notice Receive Beans at the Field. The next `amount` Pods become harvestable.
+     * @dev Amount should never exceed the number of Pods that are not yet Harvestable.
+     * @param amount Amount of Beans to receive.
+     * @param data Encoded uint256 containing the index of the Field to receive the Beans.
+     */
+    function fieldReceive(uint256 amount, bytes memory data) private {
+        uint256 fieldId = abi.decode(data, (uint256));
```

Comment:
could we add a check here that `fieldId` doesn't exceed `s.fieldCount`?

Reply:
i believe this is checked automatically in 0.8 and an appropriate panic code will be thrown. an explicit check could let us set a msg and improve debugging, but i think it would be pretty clear here.

Reply:
this is checked automatically in 0.8, right? and an appropriate panic code thrown. or do you feel it could benefit from a more useful revert message?

Reply:
added
https://github.com/BeanstalkFarms/Beanstalk/pull/860/commits/d65b6bacc7e6d96105a48c759af2f5222b0ccb26#diff-6c2813749618ebaa2ded478935db44e4d09054278e07e6ed5a595a1a6f7ab178R75